lib/agent.js

```diff
@@ -0,0 +1,98 @@
+'use strict'
+/* global WeakRef, FinalizationRegistry */
+
+const { InvalidArgumentError } = require('./core/errors')
+const { Pool } = require('../')
+
+function weakCache (fn) {
+  const cache = new Map()
+  const cleanup = new FinalizationRegistry(key => {
+    // get the WeakRef from the cache
+    const ref = cache.get(key)
+    // if the WeakRef exists and the object has been reclaimed
+    if (ref !== undefined && ref.deref() === undefined) {
+      // remove the WeakRef from the cache
+      cache.delete(key)
+    }
+  })
+  return key => {
+    // check the cache for an existing WeakRef
+    const ref = cache.get(key)
+
+    // if one exists in the cache try to return the WeakRef
+    if (ref !== undefined) {
+      const cached = ref.deref()
+      if (cached !== undefined) {
+        return cached
+      }
+    }
+
+    // otherwise, if it isn't in the cache or the reference has been cleaned up, create a new one!
+    const value = fn(key)
+    // add a WeakRef of the value to the cache
+    cache.set(key, new WeakRef(value))
+    // add the value to the finalization registry
+    cleanup.register(value, key)
+    return value
+  }
+}
```

Comment:
Move to util?

Reply:
this is agnostic enough to have its own module but for now i'll move it to util